File: src/OMS.Core/Models/Order.cs
using OMS.Core.Enums;

namespace OMS.Core.Models;

/// <summary>
/// 発注モデル
/// </summary>
public class Order
{
    /// <summary>
    /// 発注ID
    /// </summary>
    public long OrderId { get; set; }

    /// <summary>
    /// ユーザーID
    /// </summary>
    public long UserId { get; set; }

    /// <summary>
    /// 銘柄ID
    /// </summary>
    public long SecurityId { get; set; }

    /// <summary>
    /// 注文番号
    /// </summary>
    public string OrderNo { get; set; } = string.Empty;

    /// <summary>
    /// 売買区分
    /// </summary>
    public OrderSide Side { get; set; }

    /// <summary>
    /// 注文タイプ
    /// </summary>
    public OrderType OrderType { get; set; }

    /// <summary>
    /// 数量
    /// </summary>
    public decimal Quantity { get; set; }

    /// <summary>
    /// 価格（指値の場合）
    /// </summary>
    public decimal? Price { get; set; }

    /// <summary>
    /// 逆指値価格
    /// </summary>
    public decimal? StopPrice { get; set; }

    /// <summary>
    /// 有効期限
    /// </summary>
    public TimeInForce TimeInForce { get; set; }

    /// <summary>
    /// ステータス
    /// </summary>
    public OrderStatus Status { get; set; }

    /// <summary>
    /// 約定済み数量
    /// </summary>
    public decimal FilledQuantity { get; set; }

    /// <summary>
    /// 平均約定価格
    /// </summary>
    public decimal? AveragePrice { get; set; }

    /// <summary>
    /// 手数料
    /// </summary>
    public decimal? Commission { get; set; }

    /// <summary>
    /// 発注日時
    /// </summary>
    public DateTime OrderDate { get; set; }

    /// <summary>
    /// 有効期限日時
    /// </summary>
    public DateTime? ValidUntil { get; set; }

    /// <summary>
    /// 備考
    /// </summary>
    public string? Notes { get; set; }

    /// <summary>
    /// 作成日時
    /// </summary>
    public DateTime CreatedAt { get; set; }

    /// <summary>
    /// 更新日時
    /// </summary>
    public DateTime UpdatedAt { get; set; }

    // ナビゲーションプロパティ
    public Security? Security { get; set; }
    public User? User { get; set; }
    public ICollection<Execution>? Executions { get; set; }

    /// <summary>
    /// 残数量を計算
    /// </summary>
    public decimal RemainingQuantity => Quantity - FilledQuantity;

    /// <summary>
    /// 約定率を計算（%）
    /// </summary>
    public decimal FillRate => Quantity > 0 ? (FilledQuantity / Quantity) * 100 : 0;

    /// <summary>
    /// 注文が完全に約定したかどうか
    /// </summary>
    public bool IsFullyFilled => Status == OrderStatus.FILLED;

    /// <summary>
    /// 注文がアクティブかどうか
    /// </summary>
    public bool IsActive => Status == OrderStatus.NEW || Status == OrderStatus.PARTIAL;
}

File: backend/oms-api/src/main/java/com/oms/api/model/Order.java
package com.oms.api.model;

import jakarta.persistence.*;
import lombok.Data;
import java.math.BigDecimal;
import java.time.LocalDateTime;

/**
 * 発注エンティティ
 */
@Entity
@Table(name = "orders")
@Data
public class Order {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "order_id")
    private Long orderId;

    @Column(name = "user_id", nullable = false)
    private Long userId;

    @Column(name = "security_id", nullable = false)
    private Long securityId;

    @Column(name = "order_no", unique = true, nullable = false, length = 50)
    private String orderNo;

    @Enumerated(EnumType.STRING)
    @Column(name = "side", nullable = false)
    private OrderSide side;

    @Enumerated(EnumType.STRING)
    @Column(name = "order_type", nullable = false)
    private OrderType orderType;

    @Column(name = "quantity", precision = 18, scale = 4, nullable = false)
    private BigDecimal quantity;

    @Column(name = "price", precision = 18, scale = 4)
    private BigDecimal price;

    @Column(name = "stop_price", precision = 18, scale = 4)
    private BigDecimal stopPrice;

    @Enumerated(EnumType.STRING)
    @Column(name = "time_in_force", nullable = false)
    private TimeInForce timeInForce = TimeInForce.DAY;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    private OrderStatus status = OrderStatus.NEW;

    @Column(name = "filled_quantity", precision = 18, scale = 4, nullable = false)
    private BigDecimal filledQuantity = BigDecimal.ZERO;

    @Column(name = "average_price", precision = 18, scale = 4)
    private BigDecimal averagePrice;

    @Column(name = "commission", precision = 18, scale = 4)
    private BigDecimal commission;

    @Column(name = "order_date", nullable = false)
    private LocalDateTime orderDate;

    @Column(name = "valid_until")
    private LocalDateTime validUntil;

    @Column(name = "notes", columnDefinition = "TEXT")
    private String notes;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
        orderDate = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    // Transient計算プロパティ
    @Transient
    public BigDecimal getRemainingQuantity() {
        return quantity.subtract(filledQuantity);
    }

    @Transient
    public BigDecimal getFillRate() {
        if (quantity.compareTo(BigDecimal.ZERO) == 0) {
            return BigDecimal.ZERO;
        }
        return filledQuantity.divide(quantity, 4, BigDecimal.ROUND_HALF_UP)
                .multiply(new BigDecimal("100"));
    }
}
